Focus Node Concept Update

danieldresser-ie opened the pull request:
This takes the experiment I was doing on top of Tom's work, and brings it onto current master.

Some of the UI changes I made we're now pretty confident in ( like not using the border for selection ), so I've gone ahead and simplified things a little, and made renderNodeFocusRegion call the new renderFrameInternal ( and made the handling of oval nodes and dots a bit prettier, though currently with an ugly special case in the code for dots ).

I also took one last swing at an icon ... I think we're probably going to end up with a simple circle based on what I remember of previous user testing, but I do find an icon a bit clearer to describe ( if the docs say "click on the eye icon" it's really clear what they mean ), so I thought I'd have a quick try.

As for performance, one loss of performance was due to a hasLayer check that was a blacklist, so we're checking all nodules for the new render layer which they don't use.  1647098 fixes this.

On my heavy node test, current master gets 12 FPS, before the fix this got 9 FPS, after the fix it gets 10.5 FPS.  So, this is still a loss compared to master, but my hacking around in #4257 appears to show that if we change the rendering code a bit, this shouldn't be an issue.  ( I got to 14 FPS while using this PR while playing with that ).

Next step is probably another round of user testing.

danieldresser-ie commented:
Update a couple things.  Basically ready for final testing / UI approval now.  I think I'm moderately happy with the icons, though the outline should probably be a little bit thicker on the icon when it's not filled in

danieldresser-ie commented:
Now updated on top of the performance PR.  User testing has turned up some requests for UI tweaks, so there will be further changes, but this is getting pretty close hopefully.

danieldresser-ie commented:
Added a whole bunch of extra commits to this PR for UI tweaks which are currently being tested with users

danieldresser-ie commented:
Rebased onto current master, and cleaned up the last couple of hacky commits a little.

danieldresser-ie commented:
I've been focused on the UI side of this, and pretty well just kept Tom's original code for the functionality.  But when working on the active visualisation stuff, I've been working with this a bit, and I'm really wondering why the focus node is stored as a set, with clients who want to be notified of changes having to connect to both `scriptNode->focus()->memberAddedSignal()` and `scriptNode->focus()->memberRemovedSignal()`.  I guess this may have been the quickest way to implement it, but given the requirement that there is only ever one focus node ( returned by scriptNode->getFocus() ), it seems like it might be clearer if we stored this as a Node pointer and added a `focusChangedSignal()`.

johnhaddon commented:
> I'm really wondering why the focus node is stored as a set, with clients who want to be notified of changes having to connect to both scriptNode->focus()->memberAddedSignal() and scriptNode->focus()->memberRemovedSignal().

The motivation was to make it easy to integrate into NodeSetEditor, which uses a set to determine which nodes are being edited. We can reconsider though...

danieldresser-ie commented:
> The motivation was to make it easy to integrate into NodeSetEditor, which uses a set to determine which nodes are being edited. We can reconsider though...

The bit that's most ugly when I was working on the activation thing is that I need to respond to changes in the focus node by launching an expensive background process.  It's a bit annoying to start a background process on memberRemovedSignal and then immediately cancel and restart it on memberAddedSignal.  I guess not really a problem, and this complexity would be worth it for things that are actually sets, but it feels weird for something that is always a single thing ( though possibly null ).

danieldresser-ie commented:
I've now got the UI to a state where Murray and Greg are happy with it.

Possibly contentious items:
* I've moved the numeric bookmark label outside the node ( Greg did really like the more visible bookmark icon when focussed )
* The scaling of the focus icon has been changed to preserve its relationship to the border.  This now means that it's back to slightly impinging on the nodules at the side when you zoom out ( if there are multiple side nodules ), but it is much less bad than when it scaled around the center, and this way you get good clearance from the nodules on many output nodes when zoomed in, and I feel like the zooming just looks better when the placement of the icon on the border stays the same as the border grows.
* I still don't like the highlighted icon colors, but I'm OK with them if everyone else is happy

I have not yet added the focusChanged signal to scriptNode ( I'll talk with you tomorrow morning about the implications of this - can we make the focus set read-only, and require people to call ScriptNode::setFocus rather than ScriptNode::focus()->addMember()? )

Other than that, I think it's time to start thinking about what this code looks like, rather than the UI.  I'm not sure whether it's easier to review with all my little recent changes as separate commits, or if I should just squash it all together for review.

danieldresser-ie commented:
OK, this is theoretically about ready to go.

I've set the outside color of the hovered icon to $brightColor, but left the inner circle unaffected by hovering - this feels like a pretty good way to make it feel like something is getting toggled by clicking on it?

There are two commits that I haven't squashed:
* main change for today where I add the private FocusSet class
* a commit where I change one of Tom's hotkeys.  This isn't a big commit, but it's just a reminder to talk about it, since I think require Shift to be held for this feature is probably a bad idea?  Rather than sticking with F, and requiring Shift to differentiate from "framing", I feel like we should probably just pick a new key?  A silly suggestion would be "O", since it looks like the icon, but I'm more just thinking whatever key is close to right Ctrl for easy access, and isn't already used?  Maybe "S" or "D"?

Also, when Greg was asking about a hotkey to focus a bookmark, we said we didn't want to do that yet, but potentially a smaller change - would it be worth selecting a node when you jump to its bookmark in the Graph Editor?  Then you could pretty quickly hit "3", then "Ctrl-<whatever key>" to focus a bookmark in two keystrokes.

Oh, one more thing I do want to do before merging - clean up the no-op color references in the SVG file again.  Not sure if I'll keep this up in the long term, but all the back-and-forth in this PR seems to have created a bunch

johnhaddon commented on include/GafferUIBindings/GadgetBinding.h:
```diff
@@ -164,7 +164,7 @@ class GadgetWrapper : public GafferBindings::GraphComponentWrapper<WrappedType>
 					if( f )
 					{
-						f( layer, GafferUI::StylePtr( const_cast<GafferUI::Style *>( style ) ) );
+						f( layer, GafferUI::StylePtr( const_cast<GafferUI::Style *>( style ) ), reason );
```
Looks like this is throwing because we haven't bound `RenderReason` to python - see CI failures.

johnhaddon reviewed (changes requested):
Thanks Daniel!

> I've set the outside color of the hovered icon to $brightColor, but left the inner circle unaffected by hovering - this feels like a pretty good way to make it feel like something is getting toggled by clicking on it?

I think I still prefer consistent highlighting of both rings, and find the toggling of the border to be sufficient feedback. Not a big deal really - maybe we just poll a few people and go with the majority?

The other bit of fiddling I tried today was to move the focus icon a bit further out from the node so that there's no overlap/underlap with the node frame. This feels pretty good to me, and avoids half my brain wanting it on top of the node (for when I'm going to click on it) and the other half wanting it under the node (for when it's just there for display). I can show you over a screen share if you're interested.

> Rather than sticking with F, and requiring Shift to differentiate from "framing", I feel like we should probably just pick a new key? 

Definitely. I've waffled a bit about this inline, but don't have any strong ideas about what it should be.

> would it be worth selecting a node when you jump to its bookmark in the Graph Editor? Then you could pretty quickly hit "3", then "Ctrl-" to focus a bookmark in two keystrokes.

I'm a bit wary of that, because of the "oh, now my selection-following viewer has just fired up a huge compute" implications.

> OK, this is theoretically about ready to go.

I've made quite a lot of comments inline, but I think the only big thing is the awkwardness of `focusBorderWidth()`. In the absence of better ideas and not wanting to drag this out much longer, I think maybe the pragmatic solution is a comment explaining the necessity and a few thoughts about alternatives?

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1216,8 +1218,11 @@ bool ViewportGadget::buttonRelease( GadgetPtr gadget, const ButtonEvent &event )
-	/// the last button press gadget, but we erased that in buttonRelease.
+	if( m_previousClickGadget )
+	{
+		return dispatchEvent( m_previousClickGadget.get(), &Gadget::buttonDoubleClickSignal, event );
```
Trying to improve the special case where a double click actually happens doesn't seem that important - the common case is single clicks, in which case we hold on m_previousClickGadget indefinitely waiting for a double that never comes.  If it's any sort of problem, it seems like we need to deal with the common case?

As it is, it feels a bit more consistent to just let m_previousClickGadget always stick around, rather than clearing it rarely.

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -90,10 +90,11 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
-			Main = 4,
-			MidFront = 8,
-			Front = 16,
+			OverBack = 2,
```
This was Tom's commit, I didn't really mind the name that much, but I'm also fine with switching to BackMidBack.

danieldresser-ie commented on src/GafferUI/DotNodeGadget.cpp:
```diff
@@ -108,12 +125,18 @@ void DotNodeGadget::doRenderLayer( Layer layer, const Style *style, RenderReason
+
+Box3f DotNodeGadget::renderBound() const
+{
+	Box3f b = bound();
+	float maxThick = 2.0f;
+	return Box3f( b.min - V3f( maxThick, maxThick, 0.f ), b.max + V3f( maxThick, maxThick, 0.f ) );
```
Oh, I guess it's true that the bound from StandardNodeGadget would work here ... it just felt like since I'm overriding doRenderLayer I should override the bound to match, but I guess there's no actual need for that in this case.

danieldresser-ie commented on src/GafferUI/AuxiliaryNodeGadget.cpp:
```diff
@@ -153,3 +153,9 @@ bool AuxiliaryNodeGadget::updateUserColor()
+
+float AuxiliaryNodeGadget::focusBorderWidth() const
+{
+	// Can't currently focus on Auxiliary nodes
```
Hmm, do we expect the Viewer to ever show anything meaningful for an auxiliary node?  I guess once we start showing active connections, then that could be useful?

Should I add a focus icon to AuxiliaryGadget?  I guess that would mean pulling FocusGadget out of StandardNodeGadget.

danieldresser-ie replied:
Or were you just suggesting that I should disable the shortcut?

danieldresser-ie commented on src/GafferUI/BackdropNodeGadget.cpp:
```diff
@@ -578,3 +578,9 @@ bool BackdropNodeGadget::updateUserColor()
+
+float BackdropNodeGadget::focusBorderWidth() const
+{
+	// Can't focus on Backdrop
```
This is different from an auxillary node in that it isn't a single compute node, and there really isn't any good way to focus it, so in this case I probably should just disable the hotkey?

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -293,7 +301,7 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
 		/// _after_ the parent gadget has been drawn.  Whenever overriding this,
 		/// you must override layerMask and renderBound() as well.
-		virtual void doRenderLayer( Layer layer, const Style *style ) const;
+		virtual void doRenderLayer( Layer layer, const Style *style, RenderReason reason ) const;
```
Sure, that seems fine.  I guess we haven't released the UI performance overhaul yet, so it's fine to make more changes now.

danieldresser-ie commented on include/GafferUI/NodeGadget.h:
```diff
@@ -102,6 +102,8 @@ class GAFFERUI_API NodeGadget : public Gadget
 
 		std::string getToolTip( const IECore::LineSegment3f &line ) const override;
 
+		virtual float focusBorderWidth() const = 0;
```
If it was annotationBoundary, would it then be kinda weird for DotNodeGadget to use it to decide how wide to draw the focus border?

danieldresser-ie commented on src/GafferUI/GraphGadget.cpp:
```diff
@@ -894,6 +902,28 @@ void GraphGadget::selectionMemberRemoved( Gaffer::Set *set, IECore::RunTimeTyped
+{
+	if( Gaffer::Node *node = runTimeCast<Gaffer::Node>( member ) )
+	{
+		if( findNodeGadget( node ) )
+		{
+			dirty( DirtyType::Render );
+		}
```
Hmm, I might have just kept this from Tom's code ... it seems like it would be fine to just call dirty on this?

danieldresser-ie commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -69,6 +75,241 @@ using namespace Imath;
+	public :
+
+		FocusGadget( StandardNodeGadget* parent )
+			:	m_parent( parent ),
```
That's true - I guess I was just thinking about avoiding some polymorphic casts, but that's not really an issue.

danieldresser-ie replied:
I guess the downside of using parent() is that then I need to wait for parenting to change, rather than doing things in the constructor?  Unlike parent(), m_parent is guaranteed to never change, since a FocusGadget is only created during a StandardNodeGadget constuctor.

johnhaddon replied:
I think you can still do everything in the constructor - it's still reasonable to pass `parent` as an argument. It just seemed fragile/redundant to have an `m_parent` variable that mirrors state we already have access to via `parent()`. It might even make sense for the FocusGadget constructor to do `parent.addChild( this )` rather than do that on the outside?

danieldresser-ie replied:
Fixed in place.

danieldresser-ie commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -69,6 +75,241 @@ using namespace Imath;
+
+		Imath::Box3f bound() const override
+		{
+			return m_parent->bound();
```
I'm trying to recall why I did this ... it should be fine to just return an empty bound here, since focusGadget doesn't participate in layout

danieldresser-ie commented:
This is full of construction sawdust right now ... the only reason I've pushed it is that I am still wrestling with the icon.  I'm fine with highlighting the center, but I'm really struggling to like the new position.  Andrew strongly disliked the new position as well ( due to feeling like it wasn't connected enough to the node you were targetting when it isn't selected, and Ivan preferred the closer position ( because the aesthetic of how it lines up with the border looks nicer ).

I started thinking about whether it would be possible to play games with rendering under/over and started playing with it ... then I looked at your fork and saw you'd already tried this, though my approach is a little different ... selection is always underneath to avoid conflict with nodules, render is only on top when hovering.  I think I kinda like it ... I dunno, I feel like I'm spending way too long on this, but the more distanced position just feels wrong to me.

danieldresser-ie commented:
Still grinding out fixes for comments.  Feels pretty straightforward, but does take time.

Main things remaining:
* updating documentation
* switching from focusBorderWidth to annotationBound
* figuring out how to deal with BackdropGadget and AuxiliaryNodeGadget ... why doesn't AuxiliaryNodeGadget inherit from StandardNodeGadget?  If it did I could just register oval metadata and override bound() and use the standardNodeGadget functionality like Dot does?
* decide on hotkeys.

The hotkey approach I would propose ( which Murray thinks is reasonable ) is that toggle fullscreen isn't a common enough operation to deserve `` ` ``.  I would use `` ` `` for "Look at Focus" and ``Ctrl - ` `` for "Set Focus".  I don't care too much about where fullscreen goes ... not sure it deserves to be on `Space`, but since we're already using `Space` for "Maximize Pane", maybe use `Ctrl-Space`?

In the meantime, I have uploaded where I'm at if you want to look at the new icon position.

johnhaddon commented:
> switching from focusBorderWidth to annotationBound

I don't think this is essential. We already know it's a bit of a kludge, so the important thing is documenting its reason for existence and outlining how we think we could get rid of it.

> why doesn't AuxiliaryNodeGadget inherit from StandardNodeGadget?

That's a good question, and I don't recall the answer. Probably just because we know it doesn't ever want nodules, so we didn't want to pay for the empty containers?

> The hotkey approach I would propose ( which Murray thinks is reasonable ) is that toggle fullscreen isn't a common enough operation to deserve `. I would use ` for "Look at Focus" and Ctrl - ` for "Set Focus".

Agreed about fullscreen - I think `F11` might be the more standard hotkey for that anyway? But I'm not totally convinced about using `` ` `` for focus - on a UK keyboard it's at the top left rather than being in the happy lower left zone. Happy to give it a go though...

> I have uploaded where I'm at if you want to look at the new icon position.

Works for me! Thanks!

danieldresser-ie commented:
I think I've now addressed all feedback other than updating documentation and Changelog, but I haven't yet gone through and labelled which commit addresses which comment - for the number of commits in flight until I squash, that's actually going to take a bit, so you might want to wait until I do that.

johnhaddon commented:
Oh, should also have said that I played with this a bit today and having the <kbd>`</kbd> hotkey next to the numeric bookmark hotkeys seemed to work pretty well for me...

danieldresser-ie commented on resources/graphics.py:
```diff
@@ -314,8 +314,6 @@
 		'minus',
 		'navigationArrow',
 		'nodeSetDriverNodeSelection',
-		'nodeSetDriverNodeSet',
```
Fixed in place.

danieldresser-ie commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -167,11 +167,11 @@ class GAFFERUI_API ViewportGadget : public Gadget
 		/// If dragging is true, then the gadgets will be tested in DragSelect mode.
 		/// \todo Would it be more convenient for this and the space conversion functions below
 		/// to use V3fs?
-		std::vector<Gadget*> gadgetsAt( const Imath::V2f &rasterPosition, bool dragging = false ) const;
```
Fixed in place.

danieldresser-ie commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1498,7 +1498,7 @@ def editorKeyPress( editor, event ) :
 			_PinningWidget.__followNodeSelection( editor )
 			return True
-		elif event.key == "F" and event.modifiers == event.Modifiers.Shift :
+		elif event.key == "QuoteLeft" :
```
Fixed in place.

danieldresser-ie commented on resources/graphics.svg:
```diff
@@ -166,18 +166,13 @@
        id="clipPath2950">
-      <path
-         sodipodi:type="arc"
+      <ellipse
+         id="circle1896"
```
Yeah, I did also observe some pixel shifts that seemed disproportionate, but I guess the new version is probably more "correct"

danieldresser-ie commented:
Addressed last round of comments in place, added a couple more fixes in the last few commits, and went through a lot of documentation.

I should probably skim through all this code myself at least once, to make sure I haven't messed anything up in all the squashes, but other than that, I just need to put together a changelog update, and this should be good to go.

johnhaddon reviewed (changes requested):
Nearly there! We seem to have this linking error on Mac now, but I think that should just be a case of adding `QtCore` to the libs for `libGafferUI` :

```
Undefined symbols for architecture x86_64:
  "QMetaObject::Connection::~Connection()", referenced from:
      (anonymous namespace)::FocusGadget::nodeMouseEntered(boost::intrusive_ptr<GafferUI::Gadget>, GafferUI::ButtonEvent const&) in StandardNodeGadget.os
```

danieldresser-ie commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -69,6 +75,241 @@ using namespace Imath;
+			}
+		}
+
+		void setOval( bool oval )
+		{
+			m_oval = oval;
+			dirty( DirtyType::Render );
+		}
+
+		bool getOval() const
+		{
+			return m_oval;
+		}
```
I never replied to this, but now that I'm ensuring that the focus gadget exactly touches the corner, m_oval is actually used.

danieldresser-ie commented on python/GafferSceneUI/HierarchyView.py:
```diff
@@ -109,6 +109,9 @@ def __init__( self, scriptNode, **kw ) :
 			self.__pathListing.keyPressSignal().connect( Gaffer.WeakMethod( self.__keyPressSignal ), scoped = False )
 
 		self.__plug = None
+		self.setNodeSet( self.scriptNode().focusSet() )
+		# We need a manual call to _updateFromSet to initialize because NodeSetEditor may have called setNodeSet
+		# with a set with the same contents as focusSet, without emitting an _updateFromSet
```
Fixed in place.

danieldresser-ie commented on doc/source/GettingStarted/TutorialAssemblingTheGafferBot/index.md:
```diff
@@ -303,36 +307,32 @@ Congratulations! You have successfully rendered your first image. Gaffy is curre
 ## Pinning an editor to a node ##
 
-As mentioned earlier, the Viewer, Hierarchy View, and Node Editor (each an **editor**) show their respective outputs of the currently selected node. This is not always convenient, because often you will need to edit one node while viewing the output of another. You can solve this by **pinning** an editor while a node is selected, which keeps that editor focused on the node.
+As mentioned earlier, the Viewer, Hierarchy View, and Node Editor (each an **editor**) by default show their respective outputs of the current focus node. This is not always convenient, because often you will need to edit one node while viewing the output of another. You can solve this by **pinning** an editor to a selected node, which keeps that editor focused on the node.
```
Fixed in place.

danieldresser-ie commented:
Addressed all comments, updated changes.

I put together a little Gaffer thing to diff the images, and everything looks pretty OK - there are a few places where the focus icon shows up in a place where it doesn't completely make sense ... if the script sets focus on a node to show the result of an early stage, then adds more nodes, and shows the node graph, then having focus set in the middle of the new graph is a little weird, but isn't a problem - anywhere where the focus needs to be set in order to show the right result in the viewer, it seems to be getting set.  The main oddity I found from diffing the images for this PR is that `_images/parameters.png` seems to be consistently coming out slightly taller, but that's not a problem, it's just different.

The bigger problem I noticed when diffing the images is that there are some images in the docs that try to show specific states of the hierarchy view, but the sorting of the hierarchy view is random.  That doesn't seem good, but it's not related to this PR.

johnhaddon commented:
> The bigger problem I noticed when diffing the images is that there are some images in the docs that try to show specific states of the hierarchy view, but the sorting of the hierarchy view is random.

Is this showing the robot? I don't think the HierarchyView is responsible - it should always use the order that `childNames` gives, and I _think_ all nodes give consistent output there? I suspect the issue is that the `.scc` format doesn't load in a consistent ordering from run to run. Perhaps the easiest thing is to update the asset to `.abc` or `.usd`, since that's what most users will be familiar with anyway?

danieldresser-ie commented:
Implemented your suggestion of an "auxiliary" argument to a protected StandardNodeGadget constructor.  Little bit ugly, but I have added some tests that seem to show it's OK.  I've rolled it into 40272d7, since it makes sense there, and that commit was pretty simple to start with.  So, maybe this is finally done?

danieldresser-ie commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -548,12 +872,12 @@ LinearContainer *StandardNodeGadget::paddingRow()
 	else
 	{
-		return getChild<Gadget>( 0 ) // column
+		return getChild<Gadget>( 1 ) // column
```
I kind of hate this code, especially now that FocusGadget adds itself as a child, so searching for addChild in StandardNodeGadget won't even reveal why `column` isn't the first child.

johnhaddon replied:
Yeah, that and the `m_auxiliary` stuff have definitely made the internals of StandardNodeGadget harder to understand. I'm hopeful we can refactor our way out of that in the future though...
